src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/LanguageServerTarget.cs

```diff
@@ -0,0 +1,427 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.Contracts;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using StreamJsonRpc;
+
+namespace CommonLanguageServerProtocol.Framework;
+
+public interface ILspServiceProvider
+{
+    ILspServices CreateServices(string serverKind);
+}
+
+public interface ILspServices : IDisposable
+{
+    T GetRequiredService<T>();
+}
+
+public interface IRequestExecutionQueue<RequestContextType> where RequestContextType : struct
+{
+    Task<RequestContextType?> CreateRequestContextAsync(IQueueItem<RequestContextType> queueItem, CancellationToken cancellationToken);
+
+    event EventHandler<RequestShutdownEventArgs>? RequestServerShutdown;
+
+    Task<TResponseType> ExecuteAsync<TRequestType, TResponseType>(
+        bool mutatesSolutionState,
+        bool requiresLSPSolution,
+        IRequestHandler<TRequestType, TResponseType, RequestContextType> handler,
+        TRequestType request,
+        ClientCapabilities clientCapabilities,
+        string methodName,
+        CancellationToken cancellationToken);
+
+    void Shutdown();
+}
+
+public interface IQueueItem<RequestContextType> where RequestContextType : struct
```

Comment:
See top level comment - I worry about exposing this in the current form.  Its definitely leaking roslyn impl details that I don't think should be part of a public API library (for example, solution and solution state, request metrics).

I'm also having a hard time coming up with scenarios where people would want to customize this.  I get some areas (for example providing LSP services via different dependency injection method).